### PR TITLE
[MPS] Migrate SDPA from MPSGraph to Metal

### PR DESCRIPTION
**Pull request by kurtamohler**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #181030

Also fixes input broadcasting behavior.

Claude was used to co-author the `mm_simdgroup` and `softmax_rows`
functions of the Metal kernel.

Performance was measured using the following script:

<details><summary>Click to expand</summary>

```python
import torch
import torch.utils.benchmark as benchmark
import itertools

cases = [
    # (batch, heads, E, Ev, L, S, num_iters)
    (1, 8, 32, 32, 128, 128, 1000),
    (4, 12, 32, 32, 256, 256, 300),
    (8, 16, 32, 32, 512, 512, 10),
    (16, 16, 32, 32, 1024, 1024, 5),

    (1, 8, 64, 64, 128, 128, 1000),
    (4, 12, 64, 64, 256, 256, 300),
    (8, 16, 64, 64, 512, 512, 10),
    (16, 16, 64, 64, 1024, 1024, 5),

    (1, 8, 128, 128, 128, 128, 1000),
    (4, 12, 128, 128, 256, 256, 300),
    (8, 16, 128, 128, 512, 512, 10),
    (16, 16, 128, 128, 1024, 1024, 5),


    (8, 8, 64, 64, 1024, 1024, 3),

    # S doesn't scale well, because it is serialized in softmax
    (8, 8, 64, 64, 1024, 2 * 1024, 3),
    (8, 8, 64, 64, 1024, 4 * 1024, 3),
    (8, 8, 64, 64, 1024, 8 * 1024, 3),

    # L scales well
    (8, 8, 64, 64, 2 * 1024, 1024, 3),
    (8, 8, 64, 64, 4 * 1024, 1024, 3),
    (8, 8, 64, 64, 8 * 1024, 1024, 3),

    # Ev doesn't scale well, because it is iterated serially in mm
    (8, 8, 64, 2 * 64, 1024, 1024, 3),
    (8, 8, 64, 4 * 64, 1024, 1024, 3),
    (8, 8, 64, 8 * 64, 1024, 1024, 3),

    # E doesn't scale well, because it is iterated serially in mm
    (8, 8, 2 * 64, 64, 1024, 1024, 3),
    (8, 8, 4 * 64, 64, 1024, 1024, 3),
    (8, 8, 8 * 64, 64, 1024, 1024, 3),
]

dtypes = [torch.float]

def run_benchmark(device):
    print(f'dtype, batch, heads, E, Ev, L, S, baseline-time, metal-time, speedup')
    for dtype, (batch, heads, E, Ev, L, S, num_iters) in itertools.product(dtypes, cases):
        q = torch.randn(batch, heads, L, E, dtype=dtype, device='mps')
        k = torch.randn(batch, heads, S, E, dtype=dtype, device='mps')
        v = torch.randn(batch, heads, S, Ev, dtype=dtype, device='mps')

        t = benchmark.Timer(
            stmt="torch.nn.functional.scaled_dot_product_attention(q, k, v)",
            globals={"torch": torch, "q": q, "k": k, "v": v},
            num_threads=torch.get_num_threads(),
        )

        for _ in range(2):
            result_metal = t.timeit(num_iters).median

        q, k, v = (x.requires_grad_() for x in [q, k, v])

        t = benchmark.Timer(
            stmt="torch.nn.functional.scaled_dot_product_attention(q, k, v)",
            globals={"torch": torch, "q": q, "k": k, "v": v},
            num_threads=torch.get_num_threads(),
        )

        for _ in range(2):
            result_baseline = t.timeit(num_iters).median

        speedup = result_baseline/ result_metal

        print(f'{dtype}, {batch}, {heads}, {E}, {Ev}, {L}, {S}, {result_baseline:.2e}, {result_metal:.2e}, {speedup:.2f}')

run_benchmark("mps")

```

</details>

Results for float inputs when run on an Apple M4 are below. "math" refers to the implementation that just calls the higher level `aten` ops, which is currently run if the inputs require grad.

batch | heads | E | Ev | L | S | math time, s | MPSGraph time, s | Metal time, s | speedup, metal vs math | speedup, metal vs MPSGraph
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | 8 | 32 | 32 | 128 | 128 | 1.11E-04 | 5.06E-05 | 4.38E-05 | 2.53 | 1.16
4 | 12 | 32 | 32 | 256 | 256 | 1.38E-03 | 1.28E-03 | 9.10E-04 | 1.52 | 1.41
8 | 16 | 32 | 32 | 512 | 512 | 1.45E-02 | 1.32E-02 | 1.02E-02 | 1.42 | 1.29
16 | 16 | 32 | 32 | 1024 | 1024 | 1.13E-01 | 1.03E-01 | 8.08E-02 | 1.40 | 1.27
1 | 8 | 64 | 64 | 128 | 128 | 1.13E-04 | 5.39E-05 | 6.65E-05 | 1.70 | 0.81
4 | 12 | 64 | 64 | 256 | 256 | 1.54E-03 | 1.36E-03 | 1.40E-03 | 1.10 | 0.97
8 | 16 | 64 | 64 | 512 | 512 | 1.54E-02 | 1.37E-02 | 1.51E-02 | 1.02 | 0.91
16 | 16 | 64 | 64 | 1024 | 1024 | 1.17E-01 | 1.07E-01 | 1.25E-01 | 0.94 | 0.86
1 | 8 | 128 | 128 | 128 | 128 | 1.14E-04 | 6.56E-05 | 1.19E-04 | 0.96 | 0.55
4 | 12 | 128 | 128 | 256 | 256 | 1.87E-03 | 1.56E-03 | 2.42E-03 | 0.77 | 0.64
8 | 16 | 128 | 128 | 512 | 512 | 1.78E-02 | 1.55E-02 | 2.56E-02 | 0.70 | 0.61
16 | 16 | 128 | 128 | 1024 | 1024 | 1.34E-01 | 1.22E-01 | 2.29E-01 | 0.59 | 0.53
8 | 8 | 64 | 64 | 1024 | 1024 | 2.95E-02 | 2.69E-02 | 3.34E-02 | 0.88 | 0.81
8 | 8 | 64 | 64 | 1024 | 2048 | 5.84E-02 | 5.38E-02 | 7.40E-02 | 0.79 | 0.73
8 | 8 | 64 | 64 | 1024 | 4096 | 1.22E-01 | 1.13E-01 | 1.61E-01 | 0.76 | 0.70
8 | 8 | 64 | 64 | 1024 | 8192 | 2.64E-01 | 2.45E-01 | 3.41E-01 | 0.77 | 0.72
8 | 8 | 64 | 64 | 2048 | 1024 | 5.93E-02 | 5.36E-02 | 6.59E-02 | 0.90 | 0.81
8 | 8 | 64 | 64 | 4096 | 1024 | 1.16E-01 | 1.23E-01 | 1.32E-01 | 0.88 | 0.93
8 | 8 | 64 | 64 | 8192 | 1024 | 2.31E-01 | 2.11E-01 | 2.64E-01 | 0.88 | 0.80
8 | 8 | 64 | 128 | 1024 | 1024 | 3.13E-02 | 2.87E-02 | 4.72E-02 | 0.66 | 0.61
8 | 8 | 64 | 256 | 1024 | 1024 | 3.67E-02 | 3.38E-02 | 7.51E-02 | 0.49 | 0.45
8 | 8 | 64 | 512 | 1024 | 1024 | 4.63E-02 | 4.38E-02 | 1.32E-01 | 0.35 | 0.33
8 | 8 | 128 | 64 | 1024 | 1024 | 3.97E-02 | 2.88E-02 | 4.62E-02 | 0.86 | 0.62
8 | 8 | 256 | 64 | 1024 | 1024 | 3.83E-02 | 3.38E-02 | 7.52E-02 | 0.51 | 0.45
8 | 8 | 512 | 64 | 1024 | 1024 | 5.15E-02 | 4.43E-02 | 1.28E-01 | 0.40 | 0.35

### CODE REVIEW

**Comment by kurtamohler** [comment]
In this implementation, the `E`, `Ev` and `S` sizes don't scale well. I'm looking into ways to fix that

**Comment by kurtamohler** [comment]
There are a few avenues I've looked at to improve the scaling issues with this PR. I think Option 1 below is a promising way to at least approximately match or exceed the old MPSGraph impl's performance, but it will take some more work. So I figured it was worth considering some alternatives. Option 2 would likely beat option 1's performance, but would take me more time to do.

### Option 1

The `E` and `Ev` scaling issue in this PR is a problem with the matmul part of the kernel. Apple's guide for [matmul performance](https://developer.apple.com/download/files/Metal-Performance-Primitives-Programming-Guide.pdf) mentions executing multiple threadgroups per matrix. This PR's implementation is only executing one threadgroup per output matrix, so I imagine this change should improve scaling significantly. (There are also a lot of other ideas in that document like the ideal order in which a threadgroup should iterate over each of its tiles, and the ideal number of simdgroup tiles per threadgroup tile, so I could also try those out if we need.)

The `S` scaling is a problem with the softmax part of the kernel. This PR's implementation of softmax again only executes one threadgroup per output matrix, but it would be much better to execute one threadgroup for each row of each output matrix for the softmax part, like we would normally do for a reduction op. We could maybe get even better performance with loop unrolling and simd reduction.

In order to make these changes, executing more than one threadgroup per output matrix would require the op to be broken up into three kernel executions, rather than just one. The entire output matrix needs to be computed at each step (matmul, softmax, matmul) before moving on to the next, and there is no device-side way to sync multiple threadgroups. So the Metal impl would be fairly similar to the "math" backend impl (`_scaled_dot_product_attention_math`), except that the `attn_mask` behavior would likely be more performant, since we would be able to avoid the extra kernel executions that the math backend uses to apply the mask.

### Option 2

Implement the real [flash attention algorithm](https://arxiv.org/pdf/2205.14135) instead. According to the authors of the paper, the main difference between flash attention and the more mathematically straightforward algorithm is that flash attention does not write the intermediate `attn` matrix (used by the backward pass) to memory, and instead offers a fast way to recompute it in the backward pass. This is supposed to give a significant speedup, even for the backward pass, at least in CUDA, so it may be worth looking into doing this for Metal. It would take me more time to do this though, since I don't yet understand the algorithm. I'll read through the paper to try to get a basic sense about how to implement it.

There are a couple existing MIT-licensed impl of flash attention in Metal as well:
<https://github.com/philipturner/metal-flash-attention>
<https://github.com/ZimengXiong/MetalFlashAttention>

### Option 3

We could just use the math backend impl, but update the softmax and matmul ops to use a Metal kernel. This is pretty similar to Option 1, but it gives the added benefit of migrating matmul and softmax away from MPSGraph.

### Option 4

If the other options are too time consuming, we could continue using the MPSGraph impl and try to fix the existing problems with it. Perhaps it's worth checking out what kind of performance the MPSGraph function [`scaledDotProductAttention`](<https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph/scaleddotproductattention(query:key:value:mask:scale:name:)>) gives, since we don't currently use that.

**Comment by anagnorisis2peripeteia** [comment]
Hi @kurtamohler — great work getting this off the ground. I've been working on MPS Metal kernels in #181503 (Metal softmax) and have been looking at the SDPA scaling problem you've documented.

The regressions at larger head dims and sequence lengths match what I'd expect from a single-pass tiled matmul without chunking — the S and E serialization you noted is the root cause. I think your Option 2 (FlashAttention) is the right fix, and I've been implementing the online softmax algorithm as part of the softmax PR work so the tiling approach is familiar territory.

I'm planning to open a separate PR implementing FlashAttention-style forward + backward for the full-sequence case (qLen > 8), which should address the scaling regressions. I'll reference this PR in the description. Happy to coordinate if there's overlap with any direction you're heading, or if you'd prefer to take the FlashAttention approach yourself I can contribute to your branch instead.

**Comment by kurtamohler** [comment]
@anagnorisis2peripeteia, I've been studying the FlashAttention paper, and I feel that I have a fairly good grasp on it, so I'd like to try implementing it
